packages/asset-swapper/contracts/src/TwoHopSampler.sol

```diff
@@ -27,96 +27,100 @@ contract TwoHopSampler {
     using LibBytesV06 for bytes;
 
     struct HopInfo {
-        uint256 sourceIndex;
-        bytes returnData;
+        address to;
+        bytes data;
+    }
+
+    struct TwoHopResult {
+        uint256 outputAmount;
+        uint256 firstHopIndex;
+        bytes firstHopResult;
+        uint256 secondHopIndex;
+        bytes secondHopResult;
     }
 
     function sampleTwoHopSell(
-        bytes[] memory firstHopCalls,
-        bytes[] memory secondHopCalls,
+        HopInfo[] memory firstHopCalls,
+        HopInfo[] memory secondHopCalls,
         uint256 sellAmount
     )
         public
-        returns (
-            HopInfo memory firstHop,
-            HopInfo memory secondHop,
-            uint256 buyAmount
-        )
+        returns (TwoHopResult memory result)
     {
         uint256 intermediateAssetAmount = 0;
         for (uint256 i = 0; i != firstHopCalls.length; ++i) {
-            firstHopCalls[i].writeUint256(firstHopCalls[i].length - 32, sellAmount);
-            (bool didSucceed, bytes memory returnData) = address(this).call(firstHopCalls[i]);
+            bytes memory data = firstHopCalls[i].data;
+            data.writeUint256(data.length - 32, sellAmount);
```

Comment:
I forgot what a completely insane and fragile hack this two-hop stuff was. I'm thinking every sampler should expose a new function like `sampleMultiHopSell(bytes arbitraryData, uint256 takerAmount)` so we aren't making the huge assumption that they will always have amounts encoded as the final word.